### PR TITLE
Added CLI options to output server URIs to file.

### PR DESCRIPTION
**Pull request by mark-terry**
Added a CLI option to write server URIs to a specified directory. The use case for this change is to enable an external process to locate running Tessera servers, without requiring foreknowledge of the Tessera configuration.

### CODE REVIEW

**Comment by mark-terry** [comment]
Still needs tests.

File: cli/config-cli/src/main/java/com/quorum/tessera/config/cli/TesseraCommand.java
Comment (pinges) [inline cli/config-cli/src/main/java/com/quorum/tessera/config/cli/TesseraCommand.java]: Would it not be enough to just set the output path? Setting the path could mean, yes I do want the path, and here is the path?

**Review by pinges — COMMENTED** [review]
just not sure whether we need two CLI options to do that ...
Otherwise LGTM.

File: tessera-context/src/main/java/com/quorum/tessera/context/RuntimeContext.java
Comment (namtruong) [inline tessera-context/src/main/java/com/quorum/tessera/context/RuntimeContext.java]: These 2 properties (arguably only 1 is necessary) might not be needed in the RuntimeContext

File: tessera-dist/src/main/java/com/quorum/tessera/launcher/Launcher.java
Comment (namtruong) [inline tessera-dist/src/main/java/com/quorum/tessera/launcher/Launcher.java]: We might want to avoid adding logics into the Launcher as these are hidden and not covered by unit tests

File: tessera-context/src/main/java/com/quorum/tessera/context/internal/DefaultRuntimeContext.java
Comment (namtruong) [inline tessera-context/src/main/java/com/quorum/tessera/context/internal/DefaultRuntimeContext.java:138]: This might never be called

**Review by namtruong — DISMISSED** [review]
This can be done within the TesseraCommand rather in Launcher - using the Mixin pattern (similar to `PidFileMixin`)

**Review by macfarla — DISMISSED** [review]
We need this to make testing easy and safe and reliable

**Comment by Krish1979** [comment]
I am not fully convinced about the use case and Tessera is not designed to grab any port available. I discussed this with Tony and he agrees with my comment. Please discuss with Tony before proceeding further work on this change.

**Comment by macfarla** [comment]
> I am not fully convinced about the use case and Tessera is not designed to grab any port available. I discussed this with Tony and he agrees with my comment. Please discuss with Tony before proceeding further work on this change.

@Krish1979 who is Tony?

**Comment by macfarla** [comment]
> I am not fully convinced about the use case and Tessera is not designed to grab any port available.

@Krish1979 Tessera will claim a free port only when the port is not specified in the config

**Comment by macfarla** [comment]
> This can be done within the TesseraCommand rather in Launcher - using the Mixin pattern (similar to `PidFileMixin`)

@mark-terry confirming you have addressed this feedback? @namtruong can you confirm the requested changes were made and dismiss the "changes requested" status?

**Comment by antonydenyer** [comment]
Just wanted to clarify if this was only to be used for testing? If so can we make the `--outputServerURIs` option hidden and also maybe make it 'experimental' ie `--XoutputServerURIs`.

**Review by namtruong — DISMISSED** [review]
If this feature is added just to aid the acceptance testing process (i.e. starting a network of Tessera nodes) - my opinion is it's not a good enough justification. A recommended alternative is try to allocate free ports, generate the relevant Tessera configs then start Tessera nodes with them.

If this is a requested requirement i.e. genuine business use case, have we considered other implications such as : 
- If Tessera does not have a predefined port, what would go into the url of `peers` list ?
- Tessera's partyinfo exchange will not send out its actual p2p URI (I assume the port in config will be 0)
- In case of restart, Tessera might have all of its ports changed. I don't think GoQuorum or Besu at the moment have the capability to cope with this.


Additionally, those points aside, any code changes need tests (unit + integration tests) and they should be written first. The build tollgate won't be passed without 100% coverage.

**Comment by Krish1979** [comment]
I raised the same point with Tony around Tessera restart,, how will Quorum reestablish connection with Tessera after restart if the port is not static,. Currently Quorum reconnects automatically but that may not be the case after this change and would expect Quorum to restart as well.

**Comment by mark-terry** [comment]
I've done some refactoring and added test coverage for the new mixin. I've made the CLI option hidden and added the experimental prefix.

Just to clarify: this PR isn't adding the ability to use dynamic ports by specifying port 0 in the config - that feature already exists. I've just added functionality to record the server URIs in a specified location.

File: cli/cli-api/src/main/java/com/quorum/tessera/cli/parsers/ServerURIOutputMixin.java
Comment (taccatisid) [inline cli/cli-api/src/main/java/com/quorum/tessera/cli/parsers/ServerURIOutputMixin.java]: Can we move the try-catch block inside the for loop so that we continue trying to clean files up even after throwing an exception on one of them?

File: cli/cli-api/src/main/java/com/quorum/tessera/cli/parsers/ServerURIOutputMixin.java
Comment (taccatisid) [inline cli/cli-api/src/main/java/com/quorum/tessera/cli/parsers/ServerURIOutputMixin.java]: Minor style issue: this code is quite repetitive. Couldn't we have a single loop of `servers`, call `writeURIFile` on every single element and just do nothing if `fileNameFromAppType` returns `null`?

I am assuming we do not need the `findFirst()` on the stream to eliminate duplicate instances of the same service.

**Comment by Krish1979** [comment]
> I've done some refactoring and added test coverage for the new mixin. I've made the CLI option hidden and added the experimental prefix.
> 
> Just to clarify: this PR isn't adding the ability to use dynamic ports by specifying port 0 in the config - that feature already exists. I've just added functionality to record the server URIs in a specified location.

The concern was never around this CLI functionality. It is more on the deliberation of unassigned port for Tessera in acceptance tests which will negate some of Tessera "features" i.e., how will a Tessera node discover another node in the network, what will be specified in peers ?? Are there any work around in the acceptance test suites which is not obvious here ?

File: cli/cli-api/src/main/java/com/quorum/tessera/cli/parsers/ServerURIOutputMixin.java
Comment (mark-terry) [reply cli/cli-api/src/main/java/com/quorum/tessera/cli/parsers/ServerURIOutputMixin.java]: Good idea.

File: cli/cli-api/src/main/java/com/quorum/tessera/cli/parsers/ServerURIOutputMixin.java
Comment (mark-terry) [reply cli/cli-api/src/main/java/com/quorum/tessera/cli/parsers/ServerURIOutputMixin.java]: Good suggestion.

**Review by macfarla — APPROVED** [review]
Tests look good. This is a hidden option so won't disrupt any existing users

**Comment by macfarla** [comment]
> If this feature is added just to aid the acceptance testing process (i.e. starting a network of Tessera nodes) - my opinion is it's not a good enough justification. A recommended alternative is try to allocate free ports, generate the relevant Tessera configs then start Tessera nodes with them.
> 
> If this is a requested requirement i.e. genuine business use case, have we considered other implications such as :
> 
> * If Tessera does not have a predefined port, what would go into the url of `peers` list ?
> * Tessera's partyinfo exchange will not send out its actual p2p URI (I assume the port in config will be 0)
> * In case of restart, Tessera might have all of its ports changed. I don't think GoQuorum or Besu at the moment have the capability to cope with this.
> 
> Additionally, those points aside, any code changes need tests (unit + integration tests) and they should be written first. The build tollgate won't be passed without 100% coverage.

^ valid questions @namtruong - however this functionality is not added by this PR. The "use a free port" functionality already exists in Tessera. Maybe we want to doc that it's not recommended for production use, or if you want to do so here are the things you would need to do - but that's separate to this PR. 

Tests have been added so the build tollgate now passes.

**Comment by namtruong** [comment]
> > If this feature is added just to aid the acceptance testing process (i.e. starting a network of Tessera nodes) - my opinion is it's not a good enough justification. A recommended alternative is try to allocate free ports, generate the relevant Tessera configs then start Tessera nodes with them.
> > If this is a requested requirement i.e. genuine business use case, have we considered other implications such as :
> > 
> > * If Tessera does not have a predefined port, what would go into the url of `peers` list ?
> > * Tessera's partyinfo exchange will not send out its actual p2p URI (I assume the port in config will be 0)
> > * In case of restart, Tessera might have all of its ports changed. I don't think GoQuorum or Besu at the moment have the capability to cope with this.
> > 
> > Additionally, those points aside, any code changes need tests (unit + integration tests) and they should be written first. The build tollgate won't be passed without 100% coverage.
> 
> ^ valid questions @namtruong - however this functionality is not added by this PR. The "use a free port" functionality already exists in Tessera. Maybe we want to doc that it's not recommended for production use, or if you want to do so here are the things you would need to do - but that's separate to this PR.
> 
> Tests have been added so the build tollgate now passes.

The functionality does not exist in Tessera and we never promote it. You might be able to start Tessera with port 0 but it will affect other Tessera features and as a result make the whole network not function properly (some examples include partyinfo exchange, influxdb monitoring, HA setup with proxy, and possibly some more if being tested thoroughly which I doubt there has been any).

The concerns are not with the functionality raised by this PR but the justification for its necessity - and the justifications mentioned have raised more problems that have not been thoroughly tested and answered

**Review by namtruong — COMMENTED** [review]
The conversation on this PR has dragged on for some time now - so will leave it to @Krish1979 and @antonydenyer to make the final decision.

My personal view on this is 

> The use case for this is running the Tessera app without predefined ports in an acceptance test environment.

is not a good use case - and the concerns raised have not been addressed/answered properly.

**Comment by mark-terry** [comment]
I've updated the description to more accurately reflect the PR iterations and clarified the use case.

**Comment by mark-terry** [comment]
Feedback taken onboard. Closing this PR due to a change in approach.

**Pull request closed without merging**
